src/hotspot/share/oops/klass.cpp

```diff
@@ -236,6 +296,175 @@ bool Klass::can_be_primary_super_slow() const {
+  }
+}
+
+void Klass::hash_insert(Klass* klass, GrowableArray<Klass*>* secondaries, uintx& bitmap) {
+  assert(bitmap != SECONDARY_SUPERS_BITMAP_FULL, "");
+
+  int dist = 0;
+  for (int slot = klass->hash_slot(); true; slot = (slot + 1) & SECONDARY_SUPERS_TABLE_MASK) {
+    Klass* existing = secondaries->at(slot);
+    assert(((bitmap >> slot) & 1) == (existing != nullptr), "mismatch");
+    if (existing == nullptr) { // no conflict
+      secondaries->at_put(slot, klass);
+      bitmap |= uintx(1) << slot;
+      assert(bitmap != SECONDARY_SUPERS_BITMAP_FULL, "");
+      return;
+    } else {
+      // Use Robin Hood hashing to minimize the worst case search.
+      // Also, every permutation of the insertion sequence produces
+      // the same final Robin Hood hash table, provided that a
+      // consistent tie breaker is used.
+      int existing_dist = (slot - existing->hash_slot()) & SECONDARY_SUPERS_TABLE_MASK;
+      if (existing_dist < dist
+          // This tie breaker ensures that the hash order is maintained.
+          || ((existing_dist == dist)
+              && (uintptr_t(existing) < uintptr_t(klass)))) {
+        Klass* tmp = secondaries->at(slot);
+        secondaries->at_put(slot, klass);
+        klass = tmp;
+        dist = existing_dist;
+      }
+      ++dist;
+    }
+  }
+}
+
+Array<Klass*>* Klass::pack_secondary_supers(ClassLoaderData* loader_data,
+                                            GrowableArray<Klass*>* primaries,
+                                            GrowableArray<Klass*>* secondaries,
+                                            uintx& bitmap, TRAPS) {
+  int new_length = primaries->length() + secondaries->length();
+  Array<Klass*>* secondary_supers = MetadataFactory::new_array<Klass*>(loader_data, new_length, CHECK_NULL);
+
+  // Combine the two arrays into a metadata object to pack the array.
+  // The primaries are added in the reverse order, then the secondaries.
+  int fill_p = primaries->length();
+  for (int j = 0; j < fill_p; j++) {
+    secondary_supers->at_put(j, primaries->pop());  // add primaries in reverse order.
+  }
+  for( int j = 0; j < secondaries->length(); j++ ) {
+    secondary_supers->at_put(j+fill_p, secondaries->at(j));  // add secondaries on the end.
+  }
+#ifdef ASSERT
+  // We must not copy any null placeholders left over from bootstrap.
+  for (int j = 0; j < secondary_supers->length(); j++) {
+    assert(secondary_supers->at(j) != nullptr, "correct bootstrapping order");
+  }
+#endif
+
+  if (UseSecondarySupersTable) {
+    bitmap = hash_secondary_supers(secondary_supers, /*rewrite=*/true); // rewrites freshly allocated array
+  } else {
+    bitmap = SECONDARY_SUPERS_BITMAP_EMPTY;
+  }
+  return secondary_supers;
+}
+
+uintx Klass::compute_secondary_supers_bitmap(Array<Klass*>* secondary_supers) {
+  return hash_secondary_supers(secondary_supers, /*rewrite=*/false); // no rewrites allowed
+}
+
+uint8_t Klass::compute_home_slot(Klass* k, uintx bitmap) {
```

Comment:
This could use some comments, because it's not doing what I would expect, if this should match the asm code.  Here, we never call population_count() on the full bitmap (shift 0), unlike the asm code.

Reply:
Neither do we do that in assembly code [1] [2]. `Klass::compute_home_slot` is equivalent to what happens in MacroAssembler (modulo some micro-optimizations).


[1] macroAssembler_x86.cpp:
``` 
...
    int shift_count = Klass::SECONDARY_SUPERS_TABLE_MASK - bit;
    if (shift_count != 0) {
      salq(r_array_index, shift_count);
    } else {
      testq(r_array_index, r_array_index);
    }
...
  if (bit != 0) {
    popcntq(r_array_index, r_array_index);
```

[2] macroAssembler_aarch64.cpp:
```
  if (bit != 0) {
    shld(vtemp, vtemp, Klass::SECONDARY_SUPERS_TABLE_MASK - bit);
    cnt(vtemp, T8B, vtemp);
    addv(vtemp, T8B, vtemp);
    fmovd(r_array_index, vtemp);
```

Reply:
Are you asking about `if (hash > 0) {` check?

Reply:
No, in the asm code, when bit == SECONDARY_SUPERS_TABLE_MASK (MSB), we do popcnt on the full bitmap.

Reply:
That's one of micro-optimizations performed in assembly code. 
What's actually being computed is `home_index + 1`, but `r_array_base` is one word (-1) off, so first probe lands at `secondary_supers[home_index]`. It saves an instruction because you don't need to increment `r_array_index`, `r_array_base` adjustment [2] effectively increments the index as well.

There are some comments on assembly side to make it clear [1]. I was initially confused by all those small tweaks happening in assembly as well, so added more comments. Maybe we need more. 

`Klass::compute_home_slot()` is a canonical implementation to compute home slot. Comments there about what happens on assembly side would look to me irrelevant.

[1]
```
  // NB! r_array_index is off by 1. It is compensated by keeping r_array_base off by 1 word.
```

[2]
```
  // And adjust the array base to point to the data.
  // NB! Effectively increments current slot index by 1.
  assert(Array<Klass*>::base_offset_in_bytes() == wordSize, "");
  add(r_array_base, r_array_base, Array<Klass*>::base_offset_in_bytes());
```

Reply:
OK, I can see now that they both compute the same thing because the "home" bit must be 1.